scripts/ecoinvent_38_to_39_migration.py

```diff
@@ -0,0 +1,168 @@
+              )
+            return
+          new_provider = db.get(Process, new_refId_info[0])
+          exchange.defaultProviderId = new_provider.id
+          print('Updated the provider for "%s" to "%s" (%s) in process "%s" (%s).' % (
+            exchange.flow.name, new_provider.name, 
+            new_provider.refId, 
+            foreground_process.name, 
+            foreground_process.refId
+            ))
+    process_dao.update(foreground_process)
+  for process in db.getAll(Process):
+    if process.refId in mapping_dict:
+      old_ref_id = process.refId
+      process_dao.delete(process)
+      print('Deleted outdated Ecoinvent3.8 process "%s" (%s)' % (process.name, old_ref_id))
+      
+def update_product_system():
+  """This function reevaluates the existing product systems inside the database."""
+  
+  product_system_dao = ProductSystemDao(db)
+  DESCRIPTION_ADDITION = "\n\nThis product system was recalculated as it originates from Ecoinvent3.8!!!"
+  for product_system in db.getAll(ProductSystem):
+    p = product_system.referenceProcess
+    config = LinkingConfig()\
+      .providerLinking(ProviderLinking.PREFER_DEFAULTS)\
+      .preferredType(ProcessType.UNIT_PROCESS)
+    new_product_system = ProductSystemBuilder(MatrixCache.createLazy(db), config)\
+      .build(p)
+
+    new_product_system.name = product_system.name 
+    new_product_system.refId = product_system.refId
+    new_product_system.targetUnit = product_system.targetUnit
+    new_product_system.targetAmount = product_system.targetAmount
+    new_product_system.cutoff = product_system.cutoff 
+    new_product_system.referenceExchange = product_system.referenceExchange 
+    new_product_system.referenceProcess = product_system.referenceProcess 
+    new_product_system.targetFlowPropertyFactor = product_system.targetFlowPropertyFactor
+    if product_system.description == None:
+      new_product_system.description = DESCRIPTION_ADDITION    
+    elif not product_system.description.endswith(DESCRIPTION_ADDITION):
+      new_product_system.description = product_system.description + DESCRIPTION_ADDITION
+    else:
+      new_product_system.description = product_system.description
+    product_system_dao.insert(new_product_system)
+    product_system_dao.delete(product_system)
+   
+    print('Updated the product system "%s" (%s)' % (product_system.name, product_system.refId))
+
+def main():
+  """Call a file chooser with which the above code will be initiated."""
+  
+  #select the json file
+  file = FileChooser.open('*.json')
+  if file is None:
+      MessageDialog.openError(None, "Error", "You must select a json file")
+      return 
+  
+  #open json file  
+  with open(str(file), 'r') as read_file:
+    mapping_obj = json.load(read_file)
+
+  #check mapping file
+  for val in mapping_obj.values():
+    if not len(val) == 3: 
+      MessageDialog.openError(None, "Error", "You must select the correct json file")
+      return 
```

Comment:
No need to check all the values! That is a lot of computation for nothing.
You did not manage to do the following?
```suggestion
value = mapping_obj.values()[0]
if not len(value) == 3 or db.get(Process, value[0]) is None: 
  MessageDialog.openError(None, "Error", "You must select the correct json file")
  return 
```